src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceService.java

```diff
@@ -77,35 +56,64 @@ public void updateAttendance(long id, UpdateAttendanceDto attendanceDto) throws
                         .createDateTime(attendanceData.getCreateDateTime())
                         .caseNoteId(attendanceData.getCaseNoteId())
                         .build())
-                  .collect(Collectors.toSet());
+                .collect(Collectors.toSet());
     }
 
     public AbsentReasonsDto getAbsenceReasons() {
         return new AbsentReasonsDto(AbsentReason.getPaidReasons(), AbsentReason.getUnpaidReasons());
     }
 
+    @Transactional
+    public void createAttendance(final CreateAttendanceDto attendanceDto) {
 
-    private void applyAttendanceWorkflow(String offenderNo, Attendance attendance) {
-         final var eventOutcome = nomisEventOutcomeMapper.getEventOutcome(
-                 attendance.getAbsentReason(),
-                 attendance.isAttended(),
-                 attendance.isPaid());
+        var attendance = attendanceRepository.save(toAttendance(attendanceDto));
+        entityManager.flush();
+        applyAttendanceWorkflow(attendance);
+    }
 
-         nomisService.putAttendance(offenderNo, attendance.getEventId(), eventOutcome);
+    @Transactional
+    public void updateAttendance(long id, UpdateAttendanceDto attendanceDto) throws AttendanceNotFound {
 
-         if (attendance.getAbsentReason() != null && AbsentReason.getIepTriggers().contains(attendance.getAbsentReason())) {
-             final var caseNote = nomisService.postCaseNote(
-                     attendance.getOffenderBookingId(),
-                     "NEG",//"Negative Behaviour"
-                     "IEP_WARN", //"IEP Warning",
-                     attendance.getComments(),
-                     LocalDateTime.now());
+        final var attendance = attendanceRepository.findById(id)
+                .orElseThrow(AttendanceNotFound::new);
 
-             attendance.setCaseNoteId(caseNote.getCaseNoteId());
-         }
-     }
+        attendance.setAttended(attendanceDto.getAttended());
+        attendance.setPaid(attendanceDto.getPaid());
+        attendance.setAbsentReason(attendanceDto.getAbsentReason());
+        attendance.setComments(attendanceDto.getComments());
+
+        applyAttendanceWorkflow(attendance);
+
+        attendanceRepository.save(attendance);
```

Comment:
is there dirty checking/transactional advice which will mean that changes get persisted on commit anyway?

Reply:
Yeah, there is, but wasn't sure how to trigger it in my test so called it explicitly. What's the best practice on this?

Reply:
Not too sure - I think it's [pretty difficult](https://blog.apnic.net/2018/02/08/testing-springs-transactional/) to unit test without setting up mock transaction managers and associated cruft.

Could argue the transactionality is a cross cutting concern that doesn't need explicitly testing in a unit testing and is more the remit for integration tests.
Are you currently verifying the save?

Reply:
Yeah,  save is verified in the AttendanceServiceTes.